[OptionsResolver] Added support for nesting options definition

yceruto opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #4833
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/9995

I'd like to propose an alternative to #27251 and #18134 with a different approach.

It would allow you to create a nested options system with required options, validation (type, value) and
normalization.

<details>
 <summary><strong>Short documentation</strong></summary>

**To define a nested option, you can pass a closure as the default value of the option with an `OptionsResolver` argument:**
```php
$resolver
    ->defaults([
        'connection' => 'default',
        'database' => function (OptionsResolver $resolver) {
            $resolver
                ->setRequired(['dbname', 'host', ...])
                ->setDefaults([
                    'driver' => 'pdo_sqlite', 
                    'port' => function (Options $options) {
                        return 'pdo_mysql' === $options['driver'] ? 3306 : null,
                    }, 
                    'logging' => true,
                ])
                ->setAllowedValues('driver', ['pdo_sqlite', 'pdo_mysql'])
                ->setAllowedTypes('port', 'int')
                ->setAllowedTypes('logging', 'bool')
                // ...
        },
    ]);

$resolver->resolve(array(
    'database' => array(
        'dbname' => 'demo', 
        'host' => 'localhost', 
        'driver' => 'pdo_mysql',
    ),
));

// returns: array(
//    'connection' => 'default',
//    'database' => array(
//        'dbname' => 'demo',
//        'host' => 'localhost',
//        'driver' => 'pdo_mysql',
//        'port' => 3306,
//        'logging' => true,
//    ),
//)
```
Based on this instance, you can define the options under ``database`` and its desired default
value.

**If the default value of a child option depend on another option defined in parent level,
adds a second ``Options`` argument to the closure:**
```php
$resolver
    ->defaults([
        'profiling' => false,
        'database' => function (OptionsResolver $resolver, Options $parent) {
            $resolver
                ->setDefault('logging', $parent['profiling'])
                ->setAllowedTypes('logging', 'bool');
        },
    ])
;
```
**Access to nested options from lazy or normalize functions in parent level:**
```php
$resolver
    ->defaults([
        'version' => function (Options $options) {
            return $options['database']['server_version'];
        },
        'database' => function (OptionsResolver $resolver) {
            $resolver
                ->setDefault('server_version', 3.15)
                ->setAllowedTypes('server_version', 'numeric')
                // ...
        },
    ])
;
```
As condition, for nested options you must pass an array of values to resolve it at runtime, otherwise an exception will be thrown:
```php
$resolver->resolve(); // OK
$resolver->resolve(['database' => []]); // OK
$resolver->resolve(['database' => null); // KO (Exception!)
```
</details>

---

Demo app https://github.com/yceruto/nested-optionsresolver-demo

yceruto committed:
Added support for nested options definition

Doctrs commented:
What about access to children data or parent data in lazy options?
And as I see we cant add normalize or set required for nested options after defined main data

yceruto commented:
> What about access to children data or parent data in lazy options?

Access to children data in lazy options is already covered [here](https://github.com/symfony/symfony/pull/27291/files#diff-b9a7faded6c8bfc8c16c92999cc982f2R1818) 

For the rest you're right :+1: I'm thinking in another solution to allow that, so

Status: Needs Work

yceruto commented:
> What about access to ... parent data in lazy options?

Done! :tada:

> And as I see we cant add normalize or set required for nested options after defined main data

Done as well :tada: the strategy has changed a bit, now instead of passing an instance of `OptionsResolver` as default value, you can pass a closure with `OptionsResolver` argument, thus subdefinitions can alter the previous definition.

Status: Needs Review

yceruto commented:
Status: Needs Review

nicolas-grekas commented:
Does this replace #27251 completly?

yceruto commented:
PR update:
 * Added `isNested` method to know whether an option is nested or not.
 * Updated PHPDoc of the `setDefault` method about how to define nested options.
 * Added and improved test cases
 * Updated PR description and added documentation PR https://github.com/symfony/symfony-docs/pull/9995.

It would be great if someone could check this, thanks.

yceruto commented:
Any update about this?

beoboo reviewed (approved):
Hey, I just reviewed your code, and it looks good to me. I also tried it in a testing project, and I find the interface clear enough to be used.

yceruto commented:
@beoboo thanks for trying it. I hope this takes enough attention to merge into 4.2. Also, there is another PR proposing the same feature (https://github.com/symfony/symfony/pull/27251), so the members/deciders should pick one of the approaches.

Cheers!

yceruto commented:
Just rebased, no changes ;)

See LDAP Connection revamping with this feature: https://github.com/yceruto/symfony/compare/nested_options_resolver...yceruto:ldap-revamping?diff=unified

I'll expose some project examples soon.

yceruto commented:
Some research in `SonataAdminBundle` where this feature can be applied:
* [AdminType.php](https://github.com/sonata-project/SonataAdminBundle/blob/9aa22d45178a923e6557bcafc2ea5e9844df2cd4/src/Form/Type/AdminType.php#L104-L110)
* [ModelAutocompleteType.php](https://github.com/sonata-project/SonataAdminBundle/blob/9aa22d45178a923e6557bcafc2ea5e9844df2cd4/src/Form/Type/ModelAutocompleteType.php#L136)

Even fixing bugs explained by https://github.com/symfony/symfony/issues/4833

yceruto commented:
Hey! You can see now a demo for this feature in https://github.com/yceruto/nested-optionsresolver-demo, just clone it and play with nested options.

fabpot commented on src/Symfony/Component/OptionsResolver/OptionsResolver.php:
```diff
@@ -37,6 +37,13 @@ class OptionsResolver implements Options
     private $defaults = array();
 
+    /**
+     * A list of closure for nested options.
```
I think this comment can be removed.

fabpot commented on src/Symfony/Component/OptionsResolver/OptionsResolver.php:
```diff
@@ -767,6 +806,32 @@ public function offsetGet($option)
+            }
+
+            // The following section must be protected from cyclic calls.
+            // BEGIN
```
Can be removed (same for END below)

fabpot replied:
Done in c01359e00e

fabpot commented:
Thank you @yceruto.

yceruto commented:
I'm very happy to see this merged for 4.2! :tada:
